### PR TITLE
Add a warmup API for warmup triggers

### LINKED ISSUE #4588: Allow runtime to be pre-warmed along with Premium Plan worker pre-warming
In the Premium Plan we allow workers to be "pre-warmed" to mitigate cold start. Currently this method does not pre-load the runtime and results in a cold start penalty when scaling out beyond your minimum pre-warmed instances while still within your minimum plan size. 

This delay is most noticeable in languages with long initialization times like CSX, JS, etc. 

The current proposals are:

1. Add an app setting to ping an HTTP endpoint and wait for a response before routing traffic to that app. This work would happen primarily in the platform side. 
2. Add a custom trigger type that is always called before traffic is routed to the app.

In both of these proposals the host would need a loopback endpoint by default that is called if the user does not specify a custom pre-warming endpoint

### PR DESCRIPTION
**Pull request by ankitkumarr**
Fixes #4588 
The other portion will be to add a small extension registering for `warmuptrigger` type ([PR](https://github.com/Azure/azure-webjobs-sdk-extensions/pull/590)).

### Summary

Adding a new API `admin/warmup` that will be called by our internal platform for Elastic Premium skus before routing requests to a container. This will let users to warmup any code / services / connections they'd like. This also helps with warming up the entire route for executing a function.

We have exposed this for all skus but it will only be called for Elastic Premium functions for now.
For local development, we need to make sure to figure out the right behavior for core-tools (do we call this API when starting up the host?).

The function has to be called `warmup` (maybe I can pick a more unique name) and the trigger has to be of type `warmuptrigger` for this to work.

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs [hunk 1]
```diff
@@ -84,6 +88,44 @@ public async Task<IActionResult> CreateOrUpdate(string name, [FromBody] Function
...
+                        State = scriptHostManager.State.ToString(),
+                        Version = ScriptHost.Version,
+                        VersionDetails = Utility.GetInformationalVersion(typeof(ScriptHost)),
+                        Id = await hostIdProvider.GetHostIdAsync(CancellationToken.None)
+                    };
+
+                    ParameterDescriptor inputParameter = function.Parameters.First(p => p.IsTrigger);
+                    Dictionary<string, object> arguments = new Dictionary<string, object>()
+                    {
+                        { inputParameter.Name, JsonConvert.SerializeObject(hostStatus) }
```
Comment (ankitkumarr) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: Sending data to functions as `string` as that seemed simplest.

Lmk if there's concerns around it. In the extension, I will add a converter for string to JObject just in case if users want that.

File: src/WebJobs.Script.WebHost/Standby/WarmUpConstants.cs [hunk 2]
```diff
@@ -9,5 +9,6 @@ public static class WarmUpConstants
     {
         public const string FunctionName = "WarmUp";
         public const string AlternateRoute = "CSharpHttpWarmup";
+        public const string WarmupTrigger = "WarmUpTrigger";
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Standby/WarmUpConstants.cs]: Likely this trigger won't include any custom programming model or inputs. If we're just looking for a trigger type that allows string/JObject to be passed, we could either promote ManualTrigger outside of the host (since it's already a concept), or perhaps have the warmup function be an HttpTrigger? This way we don't have to define a new public binding type.

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs [hunk 3]
```diff
@@ -84,6 +88,44 @@ public async Task<IActionResult> CreateOrUpdate(string name, [FromBody] Function
             }
         }
 
+        [HttpPost]
+        [Route("admin/warmup")]
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: This belongs in HostController

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs [hunk 4]
```diff
@@ -84,6 +88,44 @@ public async Task<IActionResult> CreateOrUpdate(string name, [FromBody] Function
...
+        [Authorize(Policy = PolicyNames.AdminAuthLevel)]
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: Might consider allowing AdminAuthLevelOrInternal, unless we know that the caller will be sending a site token.

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs [hunk 5]
```diff
@@ -84,6 +88,44 @@ public async Task<IActionResult> CreateOrUpdate(string name, [FromBody] Function
...
+        [RequiresRunningHost]
+        public async Task<IActionResult> Warmup([FromServices] IScriptJobHost scriptHost, [FromServices] IScriptHostManager scriptHostManager, [FromServices] IHostIdProvider hostIdProvider)
+        {
+            // The function must be called warmup and should be of type warmup trigger
+            FunctionDescriptor function = scriptHost.GetFunctionOrNull(WarmUpConstants.FunctionName);
+            if (function != null && function.Metadata.FunctionContainsInputBinding(WarmUpConstants.WarmupTrigger))
+            {
+                try
+                {
+                    var hostStatus = new HostStatus
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: nit: in HostController we're already newing one of these up and populating in the same way. Once you move this to that controller, we should reuse code.

File: src/WebJobs.Script.WebHost/Standby/WarmUpConstants.cs (hunk 2 above)
Comment (ankitkumarr) [reply src/WebJobs.Script.WebHost/Standby/WarmUpConstants.cs]: @mathewc, if the concern is for us to do the additional work, I made the PR for the warmup extension [here](https://github.com/Azure/azure-webjobs-sdk-extensions/pull/590).

But if you have other thoughts, please let us know. We can discuss offline, and if we think leveraging our existing extensions is the right approach, I would be happy to update this PR.

Comment (mathewc) [reply src/WebJobs.Script.WebHost/Standby/WarmUpConstants.cs]: It's not about the work involved to add a new trigger type, it's about increasing concept count (a new binding) with no real benefit. If we just move ManualTrigger out into the Extensions library (peer to Timer and other core built ins) user's can use that. That also has the benefit of addressing the issue that currently precompiled functions cannot use ManualTrigger.

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs [hunk 6]
```diff
@@ -84,6 +88,44 @@ public async Task<IActionResult> CreateOrUpdate(string name, [FromBody] Function
...
+                    {
+                        State = scriptHostManager.State.ToString(),
+                        Version = ScriptHost.Version,
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: How is it expected that a warmup function would use this information? I don't see how it would. Since this becomes a contract, it'd be good to only send information we think they actually need.

File: src/WebJobs.Script.WebHost/Standby/WarmUpConstants.cs (hunk 2 above)
Comment (ankitkumarr) [reply src/WebJobs.Script.WebHost/Standby/WarmUpConstants.cs]: Thanks Mathew! Yeah I think initially we didn't go with ManualTrigger as it's part of the host right now and as you said working with precompile would be a challenge.

I think your suggestion makes sense to move it to extensions repo. I am guessing, from the host side it should not be a breaking change that `ManualTrigger` with not be passed to sdk as `NoAutomaticTrigger` anymore (from my understanding that's how it works today).

I am ok with leveraging the Manual trigger. I will wait to hear @fabiocav remarks before making that change.

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs (hunk 6 above)
Comment (ankitkumarr) [reply src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: You are right in that version may not be as useful. But, I think that `State` of the function can definitely be of some use. Like say if app is offline, maybe a user would want to not initialize certain components.

I start with just adding "State" in the payload and as we find other useful properties, we can update this.

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs (hunk 5 above)
Comment (ankitkumarr) [reply src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: Now that I am only passing the hostState, there's not much to reuse.

File: src/WebJobs.Script.WebHost/Standby/WarmUpConstants.cs (hunk 2 above)
Comment (ankitkumarr) [reply src/WebJobs.Script.WebHost/Standby/WarmUpConstants.cs]: As per offline discussion, we will be going ahead with a new `WarmupTrigger`.

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs [hunk 7]
```diff
@@ -287,6 +290,33 @@ public IActionResult GetAdminToken()
...
+        [Route("admin/warmup")]
+        [Authorize(Policy = PolicyNames.AdminAuthLevelOrInternal)]
+        [RequiresRunningHost]
+        public async Task<IActionResult> Warmup([FromServices] IScriptJobHost scriptHost, [FromServices] IScriptHostManager scriptHostManager, [FromServices] IHostIdProvider hostIdProvider)
+        {
+            // The function must be called warmup and should be of type warmup trigger
+            FunctionDescriptor function = scriptHost.GetFunctionOrNull(WarmUpConstants.FunctionName);
+            if (function != null && function.Metadata.ContainsInputBinding(WarmUpConstants.WarmupTrigger))
+            {
+                var hostStatus = new HostStatus
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/HostController.cs]: I think we need a strongly typed contract here, e.g. a WarmupContext that we new up and invoke the function with. Need to decide what data to pass initially as part of this context. Not sure State is useful...

Comment (ankitkumarr) [reply src/WebJobs.Script.WebHost/Controllers/HostController.cs]: Chatted offline. I agree with a WarmupContext. But, having a hard time figuring out the right info. If anyone has suggestions, that would help.

Also wondering if something like "InstanceCount" would have made sense. This info would have to come from the caller. So we can't add that now.

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs [hunk 8]
```diff
@@ -287,6 +290,33 @@ public IActionResult GetAdminToken()
...
         }
 
+        [HttpPost]
+        [Route("admin/warmup")]
+        [Authorize(Policy = PolicyNames.AdminAuthLevelOrInternal)]
+        [RequiresRunningHost]
+        public async Task<IActionResult> Warmup([FromServices] IScriptJobHost scriptHost, [FromServices] IScriptHostManager scriptHostManager, [FromServices] IHostIdProvider hostIdProvider)
+        {
+            // The function must be called warmup and should be of type warmup trigger
+            FunctionDescriptor function = scriptHost.GetFunctionOrNull(WarmUpConstants.FunctionName);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/HostController.cs]: nit: recommend moving this lookup logic into a ScriptHost.GetWarmupFunctionOrNull helper

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs [hunk 9]
```diff
@@ -287,6 +290,33 @@ public IActionResult GetAdminToken()
...
+            {
+                var hostStatus = new HostStatus
+                {
+                    State = scriptHostManager.State.ToString()
+                };
+
+                ParameterDescriptor inputParameter = function.Parameters.First(p => p.IsTrigger);
+                var arguments = new Dictionary<string, object>()
+                {
+                    { inputParameter.Name, JsonConvert.SerializeObject(hostStatus) }
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/HostController.cs]: Once we move to WarmupContext, we won't serialize it - we'll pass the instance in directly

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs (hunk 7 above)
Comment (ankitkumarr) [reply src/WebJobs.Script.WebHost/Controllers/HostController.cs]: For now, we will have an empty `WarmupContext` object. fyi, @fabiocav

**Comment by ankitkumarr** [comment]
Once https://github.com/Azure/azure-webjobs-sdk-extensions/pull/590 is merged (and released as a nuget), I need to update this PR to use a `WarmupContext` and update the reference to the extensions.

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs (hunk 9 above)
Comment (fabiocav) [reply src/WebJobs.Script.WebHost/Controllers/HostController.cs]: Is this still planned?

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs [hunk 10]
```diff
@@ -287,6 +290,33 @@ public IActionResult GetAdminToken()
...
 
+        [HttpPost]
+        [Route("admin/warmup")]
+        [Authorize(Policy = PolicyNames.AdminAuthLevelOrInternal)]
+        [RequiresRunningHost]
+        public async Task<IActionResult> Warmup([FromServices] IScriptJobHost scriptHost, [FromServices] IScriptHostManager scriptHostManager, [FromServices] IHostIdProvider hostIdProvider)
+        {
+            // The function must be called warmup and should be of type warmup trigger
+            FunctionDescriptor function = scriptHost.GetFunctionOrNull(WarmUpConstants.FunctionName);
+            if (function != null && function.Metadata.ContainsInputBinding(WarmUpConstants.WarmupTrigger))
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Controllers/HostController.cs]: Change the extension method to filter on triggers only and rename the method to something like `HasTrigger` taking the trigger type name, or `HasTriggerType`

File: src/WebJobs.Script.WebHost/Standby/WarmUpConstants.cs [hunk 11]
```diff
@@ -9,5 +9,6 @@ public static class WarmUpConstants
     {
         public const string FunctionName = "WarmUp";
         public const string AlternateRoute = "CSharpHttpWarmup";
+        public const string WarmupTrigger = "WarmupTrigger";
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Standby/WarmUpConstants.cs]: nit: `WarmupTrigger` -> `WarmupTriggerName`

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs (hunk 8 above)
Comment (ankitkumarr) [reply src/WebJobs.Script.WebHost/Controllers/HostController.cs]: Do you think the ScriptHost should have the knowledge of `WarmupTrigger`? I thought from Script Host's point of view, it's nothing special and shouldn't have a special getter for it.

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs (hunk 9 above)
Comment (mathewc) [reply src/WebJobs.Script.WebHost/Controllers/HostController.cs]: Yes - we should remove this serialization now.

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs (hunk 8 above)
Comment (ankitkumarr) [reply src/WebJobs.Script.WebHost/Controllers/HostController.cs]: I guess if I add it in the extension, it's probably fine and cleaner that way. I will do that. Thanks for the suggestion.

File: src/WebJobs.Script/WebJobs.Script.csproj [hunk 12]
```diff
@@ -42,7 +42,7 @@
     <PackageReference Include="Microsoft.Azure.Functions.NodeJsWorker" Version="1.1.1" />
     <PackageReference Include="Microsoft.Azure.Functions.PowerShellWorker" Version="0.1.174-preview" />
     <PackageReference Include="Microsoft.Azure.WebJobs" Version="3.0.14-11656" />
-    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions" Version="3.0.3" />
+    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions" Version="3.0.4" />
```
Comment (ankitkumarr) [inline src/WebJobs.Script/WebJobs.Script.csproj]: Should this be a separate PR. Added to this for the sake of build times.

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs [hunk 13]
```diff
@@ -287,6 +290,27 @@ public IActionResult GetAdminToken()
             return Ok(requestHeaderToken);
         }
 
+        [HttpPost]
+        [Route("admin/warmup")]
+        [Authorize(Policy = PolicyNames.AdminAuthLevelOrInternal)]
+        [RequiresRunningHost]
+        public async Task<IActionResult> Warmup([FromServices] IScriptJobHost scriptHost, [FromServices] IScriptHostManager scriptHostManager, [FromServices] IHostIdProvider hostIdProvider)
+        {
+            FunctionDescriptor function = scriptHost.GetWarmupFunctionOrNull();
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/HostController.cs]: Nice. Now you could move the rest of this method into your new helper extension class, so you could simply do `scriptHost.InvokeWarmupAsync()` or perhaps `bool scriptHost.TryInvokeWarmupAsync()`. That keeps the controller method clean, and makes the logic easily testable.

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs [hunk 14]
```diff
@@ -287,6 +287,16 @@ public IActionResult GetAdminToken()
             return Ok(requestHeaderToken);
         }
 
+        [HttpPost]
```
Comment (ankitkumarr) [inline src/WebJobs.Script.WebHost/Controllers/HostController.cs]: Need to double check that this is expected to be a POST request.

Comment (ankitkumarr) [reply src/WebJobs.Script.WebHost/Controllers/HostController.cs]: Looking at the platform code, this needs to be GET request.
Phew! That have been a tiny disaster.

Comment (ankitkumarr) [reply src/WebJobs.Script.WebHost/Controllers/HostController.cs]: Let's make it GET and POST to be safe.

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs (hunk 10 above)
Comment (ankitkumarr) [reply src/WebJobs.Script.WebHost/Controllers/HostController.cs]: This is not relevant anymore as moved to another extension.

File: src/WebJobs.Script/Host/ScriptJobHostExtensions.cs [hunk 15]
```diff
@@ -19,5 +23,52 @@ public static FunctionDescriptor GetFunctionOrNull(this IScriptJobHost scriptJob
...
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Try to invoke a warmup function if available
+        /// </summary>
+        /// <returns>true if a warmup function was invoked, false otherwise</returns>
+        public static async Task<bool> TryInvokeWarmupAsync(this IScriptJobHost scriptJobHost)
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/ScriptJobHostExtensions.cs]: nit: return comment doesn't match the return (task type).

File: src/WebJobs.Script/WebJobs.Script.csproj (hunk 12 above)
Comment (fabiocav) [reply src/WebJobs.Script/WebJobs.Script.csproj]: Fine here.

File: src/WebJobs.Script/Host/ScriptJobHostExtensions.cs [hunk 16]
```diff
@@ -19,5 +23,52 @@ public static FunctionDescriptor GetFunctionOrNull(this IScriptJobHost scriptJob
...
+                    .InputBindings
+                    .Where(b => b.IsTrigger)
+                    .Any(b => b.Type.Equals(WarmupTriggerName, StringComparison.OrdinalIgnoreCase));
+
+                if (isWarmupTrigger)
+                {
+                    return function;
+                }
+            }
+
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/ScriptJobHostExtensions.cs]: Perhaps go with this approach:

``` csharp
 /// <summary>
        /// Lookup a function by name
        /// </summary>
        /// <param name="name">name of function</param>
        /// <returns>function or null if not found</returns>
        public static FunctionDescriptor GetFunctionOrNull(this IScriptJobHost scriptJobHost, string name)
        {
            return scriptJobHost.Functions.FirstOrDefault(f => IsFunctionNameMatch(f.Name, name));
        }

        public static bool IsFunctionNameMatch(string functionName, string comparison)
        {
            return string.Equals(functionName, comparison, StringComparison.OrdinalIgnoreCase);
        }

        /// <summary>
        /// Lookup a warmup function
        /// </summary>
        /// <returns>Warmup function or null if not found</returns>
        public static FunctionDescriptor GetWarmupFunctionOrNull(this IScriptJobHost scriptJobHost)
        {
            return scriptJobHost.Functions.FirstOrDefault(f =>
            {
                return IsFunctionNameMatch(f.Name, WarmupFunctionName)
                && f.Metadata
                    .InputBindings
                    .Any(b => b.IsTrigger && b.Type.Equals(WarmupTriggerName, StringComparison.OrdinalIgnoreCase));
            });
        }
```

Comment (ankitkumarr) [reply src/WebJobs.Script/Host/ScriptJobHostExtensions.cs]: Thanks Fabio! That does look a lot cleaner :)

**Comment by ankitkumarr** [comment]
Two Application Insights test are failing because the number of logs are different.
There's a new log line
"Initializing warmup extension"

I will update those tests in a little bit.

File: test/WebJobs.Script.Tests.Integration/ApplicationInsights/ApplicationInsightsEndToEndTestsBase.cs [hunk 17]
```diff
@@ -287,6 +287,7 @@ public async Task Validate_HostLogs()
             ValidateTrace(traces[idx++], "Host lock lease acquired by instance ID", ScriptConstants.LogCategoryHostGeneral);
             ValidateTrace(traces[idx++], "Host started (", LogCategories.Startup);
             ValidateTrace(traces[idx++], "Initializing Host", LogCategories.Startup);
+            ValidateTrace(traces[idx++], "Initializing Warmup Extension", LogCategories.CreateTriggerCategory("Warmup"));
```
Comment (ankitkumarr) [inline test/WebJobs.Script.Tests.Integration/ApplicationInsights/ApplicationInsightsEndToEndTestsBase.cs]: @fabiocav / @mathewc, this is being logged during host initialization (had to add here to make sure tests pass). I am not sure why it happens only for warmup, and not for timers / files which are also initialized during startup. I did not add any special logging in that extension either.

I will not block this PR as it's only a log entry. But, just want to make sure this is expected and ok?

Comment (ankitkumarr) [inline test/WebJobs.Script.Tests.Integration/ApplicationInsights/ApplicationInsightsEndToEndTestsBase.cs]: I had to do this because I had added 
https://github.com/Azure/azure-webjobs-sdk-extensions/blob/515576b4226bcf980792a3a57684e187f0be9d1a/src/WebJobs.Extensions/Extensions/Warmup/WarmupConfigProvider.cs#L34

Guessing it's not a big deal to have this log at every startup. But just wanted to make a note of it in case there's concerns.
